drivers/i2c/i2c_mchp_xec.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <clock_control.h>
+#include <kernel.h>
+#include <soc.h>
+#include <errno.h>
+#include <i2c.h>
+
+#define SPEED_100KHZ_BUS 0
+#define SPEED_400KHZ_BUS 1
+#define SPEED_1MHZ_BUS	 2
+
+#define WAIT_INTERVAL	5
+/* 100 us */
+#define TIMEOUT		20
+/* 25 us */
+#define MAX_CLK_STRETCHING	5
+
+struct xec_speed_cfg {
+	u32_t bus_clk;
+	u32_t data_timing;
+	u32_t start_hold_time;
+	u32_t config;
+	u32_t timeout_scale;
+};
+
+struct i2c_xec_config {
+	u32_t port_sel;
+	u32_t base_addr;
+	u8_t irq;
+};
+
+struct i2c_xec_data {
+	u32_t pending_stop;
+};
+
+/* Recommended programming values based on 16MHz
+ * i2c_baud_clk_period/bus_clk_period - 2 = (low_period + hi_period)
+ * bus_clk_reg (16MHz/100KHz -2) = 0x4F + 0x4F
+ *             (16MHz/400KHz -2) = 0x0F + 0x17
+ *             (16MHz/1MHz -2) = 0x05 + 0x09
+ */
+static const struct xec_speed_cfg xec_cfg_params[] = {
+	[SPEED_100KHZ_BUS] = {
+		.bus_clk            = 0x00004F4F,
+		.data_timing        = 0x0C4D5006,
+		.start_hold_time    = 0x0000004D,
+		.config             = 0x01FC01ED,
+		.timeout_scale      = 0x4B9CC2C7,
+	},
+	[SPEED_400KHZ_BUS] = {
+		.bus_clk            = 0x00000F17,
+		.data_timing        = 0x040A0A06,
+		.start_hold_time    = 0x0000000A,
+		.config             = 0x01000050,
+		.timeout_scale      = 0x159CC2C7,
+	},
+	[SPEED_1MHZ_BUS] = {
+		.bus_clk            = 0x00000509,
+		.data_timing        = 0x04060601,
+		.start_hold_time    = 0x00000006,
+		.config             = 0x10000050,
+		.timeout_scale      = 0x089CC2C7,
+	},
+};
+
+static int counter;
```

Comment:
I don't think this will work well with multiple driver instances doing things concurrently...shouldn't this be a member of device->driver_data instead?

Reply:
Agree. function now receives a pointer like other drivers handling multiple instances